akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSource.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.io
+
+import java.io.InputStream
+
+import akka.annotation.InternalApi
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.{
+  AbruptStageTerminationException,
+  Attributes,
+  IOOperationIncompleteException,
+  IOResult,
+  Outlet,
+  SourceShape
+}
+import akka.stream.stage.{ GraphStageLogic, GraphStageLogicWithLogging, GraphStageWithMaterializedValue, OutHandler }
+import akka.util.ByteString
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+case class DownstreamFinishedException()
+    extends RuntimeException("Downstream Finished before stream was fully consumed")
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class InputStreamSource(factory: () => InputStream, chunkSize: Int)
+    extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+
+  require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+
+  private val out: Outlet[ByteString] = Outlet("InputStreamSource.out")
+
+  override def shape: SourceShape[ByteString] = SourceShape(out)
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.inputStreamSource
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val mat = Promise[IOResult]
+    val logic = new GraphStageLogicWithLogging(shape) with OutHandler {
+      private val buffer = new Array[Byte](chunkSize)
+      private var readBytesTotal = 0L
+      private var inputStream: InputStream = _
+      private var closed = false
```

Comment:
`closed` seems to be pretty much a flag that is set when completing the promise, so maybe we can make `closed` a `def` checking if the promise is completed instead?